### PR TITLE
feat(hardhat): add rollup consensus service

### LINKED ISSUE #659: feat: add rollup consensus service and related setup
### Summary
Introduce functionality to support rollup consensus in the backend. This includes creating necessary files, initializing services, setting up contracts, and adding an endpoint to retrieve consensus contract addresses.

### Tasks
1. Create `/backend/rollup/consensus_service.py`:
   - Implement the core logic for handling rollup consensus operations.
2. Initialize `consensus_service` in `/backend/protocol_rpc/server.py`.
3. Setup Hardhat environment:
   - Ensure Hardhat is properly configured for deploying contracts.
4. Deploy all contracts:
   - Follow the process outlined in [this example](https://github.com/yeagerai/genlayer-consensus/blob/05474514652c41c319141a136a645aa99b6ac3f0/test/TransactionFlow.ts).
   - Include a deployment script (`deployFixture`) for initialization.
5. Add a method and corresponding API endpoint:
   - Return all consensus contract addresses.

### Acceptance Criteria
- `/backend/rollup/consensus_service.py` is created and functional.
- `consensus_service` is initialized in `/backend/protocol_rpc/server.py`.
- Hardhat is fully set up and functional for deploying contracts.
- Deployment of all contracts follows the specified process and includes the `deployFixture` script.
- API endpoint is added to return consensus contract addresses.

### Additional Notes
- Ensure all changes are covered by unit tests.
- Verify the deployment process with Hardhat.
- Update documentation if needed to include details about the new service and API endpoint.

### LINKED ISSUE #671: Display version in the studio frontend for easy reference

### PR DESCRIPTION
**Pull request by mpaya5**
Fixes #659 and #669

# What

- Created `/backend/rollup/consensus_service.py` to handle rollup consensus operations
- Added contract deployment persistence in Hardhat environment by using Hardhat Ignition
- Implemented `deployFixture` logic in Hardhat deployment
- Set up Docker volumes for deployment contracts
- Created contract loading functionality in ConsensusService

# Why

- To enable interaction with consensus contracts from the backend
- To ensure contract deployments persist between container restarts
- To maintain consistent contract initialization across environments
- To provide a robust interface for consensus operations

# Testing done

- Verified contract deployment persistence after container restarts
- Tested contract initialization through `deployFixture` logic in TransactionFlow
- Verified contract loading in ConsensusService

# Decisions made

- Implemented `deployFixture` logic in Hardhat deployment
- Implemented Hardhat Ignition for having deployment persistence
- Structured ConsensusService to load contracts individually

# Checks

- [x] I have tested this code
- [x] I have reviewed my own PR
- [x] I have created an issue for this PR
- [x] I have set a descriptive PR title compliant with conventional commits

# Reviewing tips

- Check the contract initialization order in `deployFixture`
- Verify the contract loading logic in ConsensusService
- Pay attention to the error handling in contract loading
- Review the Docker volume configurations

## Additional notes

By merging this PR, we can close https://github.com/yeagerai/genlayer-studio/pull/670 and https://github.com/yeagerai/genlayer-studio/issues/669 as this implementation already includes the persistence functionality that was planned there.

### This PR implements:

- Docker volume mappings for deployments, artifacts and cache
- Deployment logic following `deployFixture`
- Deployment persistence between container restarts

All the functionality planned for https://github.com/yeagerai/genlayer-studio/pull/670 and https://github.com/yeagerai/genlayer-studio/issues/669 is covered here as part of the larger consensus service integration.

### Dependency

This PR depends on #704

# User facing release notes

This PR introduces the consensus service backend integration with persistent contract deployments. It enables reliable interaction with consensus contracts and ensures contract deployments persist between system restarts. The service provides a robust interface for consensus operations and maintains consistent contract initialization across environments.

### CODE REVIEW

**Comment by mpaya5** [comment]
By merging this PR, we can close #670 and #669 as this implementation already includes the persistence functionality that was planned there. 

This PR implements:
- Docker volume mappings for deployments, artifacts and cache
- Manual deployment file saving
- Proper file permissions
- Deployment persistence between container restarts

All the functionality planned for #670 and #669 is covered here as part of the larger consensus service integration.

**Comment by mpaya5** [comment]
Closing this PR as the persistence functionality will be implemented as part of #713, which provides a more comprehensive solution including:

- Contract deployment persistence through Docker volumes
- Proper file handling and permissions
- Integration with the consensus service
- Complete deployment workflow

The requirements from this PR will be fully covered there as part of the larger consensus service integration. Please refer to #713 for the implementation.

**Comment by mpaya5** [comment]
We need to add more tests for the hardhat deploy functionality and solve the problems related with the Github Actions tests

File: backend/protocol_rpc/endpoints.py [hunk 1]
```diff
@@ -12,6 +12,7 @@
 
 from backend.database_handler.contract_snapshot import ContractSnapshot
 from backend.database_handler.llm_providers import LLMProviderRegistry
+from backend.rollup.consensus_service import ConsensusService
```
Comment (cristiam86) [inline backend/protocol_rpc/endpoints.py:15]: @mpaya5 this import is not being used

File: backend/rollup/consensus_service.py [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+                f"CONSENSUS_SERVICE: Error decoding {contract_name} deployment file: {str(e)}"
+            )
+            return None
+        except Exception as e:
+            print(
+                f"CONSENSUS_SERVICE: Error loading {contract_name} contract: {str(e)}"
+            )
+            return None
+
+    def _send_transaction(self, contract_function):
```
Comment (cristiam86) [inline backend/rollup/consensus_service.py:96]: @mpaya5 where is this function being used?

**Review by cristiam86 — CHANGES_REQUESTED** [review]

File: backend/protocol_rpc/endpoints.py (hunk 1 above)
Comment (mpaya5) [reply backend/protocol_rpc/endpoints.py:15]: It's used in the #746

**Review by cristiam86 — APPROVED** [review]

**Pull request merged**